src/ui/popupmenu/popupmenu.rs

```diff
@@ -442,10 +444,12 @@ impl Popupmenu {
                     *id.borrow_mut() = Some(sig_id);
                 }
 
-                // Update the `kind` icon with "selected" fg color.
-                let buf =
-                    get_icon_pixbuf(&item.item.kind, &fg_sel, font_height);
-                item.kind.set_from_pixbuf(&buf);
+                // Update the `kind` icon with "selected" fg color if it is an icon.
+                let buf = get_icon_pixbuf(&item.item.kind, &fg_sel, font_height);
+                if items.iter().any(|item| !item.kind.is_unknown()) {
```

Comment:
I wonder if we can assign this to a variable to avoid running this two times, here and on line 393?

We would have to move line 398 and the check below it before the assignment.

Just a nitpick

Reply:
Do you mean like this?
```diff
diff --git a/src/ui/popupmenu/popupmenu.rs b/src/ui/popupmenu/popupmenu.rs
index b4f5248..9425f3c 100644
--- a/src/ui/popupmenu/popupmenu.rs
+++ b/src/ui/popupmenu/popupmenu.rs
@@ -383,6 +383,7 @@ impl Popupmenu {
 
         self.items.once_loaded(Some(item_num), move |items| {
             let mut state = state.borrow_mut();
+            let items_have_kinds = items.iter().any(|item| !item.kind.is_unknown());
 
             if let Some(prev) = items.get(state.selected as usize) {
                 prev.info.set_visible(false);
@@ -390,9 +391,7 @@ impl Popupmenu {
 
                 // Update the `kind` icon with default fg color if it is an icon.
                 let buf = get_icon_pixbuf(&prev.item.kind, &fg, font_height);
-                if items.iter().any(|item| !item.kind.is_unknown())  {
-                    prev.image.set_from_pixbuf(&buf);
-                }
+                if items_have_kinds { prev.image.set_from_pixbuf(&buf) }
             }
 
             state.selected = item_num;
@@ -446,10 +445,8 @@ impl Popupmenu {
 
                 // Update the `kind` icon with "selected" fg color if it is an icon.
                 let buf = get_icon_pixbuf(&item.item.kind, &fg_sel, font_height);
-                if items.iter().any(|item| !item.kind.is_unknown()) {
-                    item.image.set_from_pixbuf(&buf);
-                }
-                
+                if items_have_kinds { item.image.set_from_pixbuf(&buf) }
+               
                 let newline =
                     if item.item.menu.len() > 0 && item.item.info.len() > 0 {
                         "\n"
```
The only problem here is it happens even if `items_have_kinds` isn't used. (Also, I am not sure on the variable name)

Reply:
The only place where it's not used is on https://github.com/vhakulinen/gnvim/blob/5bbccf3d0328cf31ed60a616f51de1a29eba492c/src/ui/popupmenu/popupmenu.rs#L398-L413

Yeah, it seems it's not a good idea to touch this, it's the code path when the popup menu is just opened, nevermind.

Reply:
So leave it as is?